WeRadiate-ThermoSense.ino

```diff
@@ -405,6 +464,214 @@ void loop(void)
+                {
+                sensor_CompostTemp.requestTemperatures();
+                float compostTempC = sensor_CompostTemp.getTempCByIndex(0);
+                Serial.print("Compost temperature: "); Serial.print(compostTempC); Serial.println(" C");
+                // transmit the measurement
+                b.putT(compostTempC);
+                flag |= FlagsSensor3::FlagWater;
+                }
+        else if (fHasCompostTemp)
+                {
+                gCatena.SafePrintf("No compost temperature\n");
+                }
+             
+        /* set D11 low to turn off after measuring */        
+//        digitalWrite(D11, LOW);
+        pinMode(D11, INPUT);
+      
+        *pFlag = uint8_t(flag);
+        if (savedLed != LedPattern::Joining)
+                {
+                gLed.Set(LedPattern::Sending);
+                }
+        else
+                {
+                gLed.Set(LedPattern::Joining);
+                }
+        
+        gLoRaWAN.SendBuffer(b.getbase(), b.getn(), sendBufferDoneCb, NULL);
+}
+
+void
+sendBufferDoneCb(
+        void *pContext,
+        bool fStatus
+        )
+        {
+        osjobcb_t pFn;
+
+        gLed.Set(LedPattern::Settling);
+        if (! fStatus)
+                {
+                gCatena.SafePrintf("send buffer failed\n");
+                pFn = txFailedDoneCb;
+                }
+        else
+                {
+                pFn = settleDoneCb;
+                }
+        os_setTimedCallback(
+                &sensorJob,
+                os_getTime()+sec2osticks(CATCFG_T_SETTLE),
+                pFn
+                );
+        }
+
+void
+txFailedDoneCb(
+        osjob_t *pSendJob
+        )
+        {
+        gCatena.SafePrintf("not provisioned, idling\n");
+        gLoRaWAN.Shutdown();
+        gLed.Set(LedPattern::NotProvisioned);
+        }
+
+void settleDoneCb(
+        osjob_t *pSendJob
+        )
+        {
+        // if connected to USB, don't sleep
+        // ditto if we're monitoring pulses.
+	// XXX: do we need this?
+#ifdef USBCON
+        if (fUsbPower || fHasPower1)
+#else
+        if (fHasPower1)
+#endif
+        {
+        gLed.Set(LedPattern::Sleeping);
+        os_setTimedCallback(
+                &sensorJob,
+                os_getTime() + sec2osticks(CATCFG_T_INTERVAL),
+                sleepDoneCb
+                );
+        return;
+        }
+
+        /* ok... now it's time for a deep sleep */
+        gLed.Set(LedPattern::Off);
+        Serial.end();
+        Wire.end();
+        SPI.end();
+	if (fFlash)
+		gSPI2.end();
+
+        gCatena.Sleep(CATCFG_T_INTERVAL);
+
+        /* and now... we're awake again. trigger another measurement */
+        Serial.begin();
+        Wire.begin();
+        SPI.begin();
+	if (fFlash)
+		gSPI2.begin();
+
+        sleepDoneCb(pSendJob);
+        }
+
+void sleepDoneCb(
+        osjob_t *pJob
+        )
+        {
+        gLed.Set(LedPattern::WarmingUp);
+        gSi1133.start();
```

Comment:
this looks suspicious. It probably should be in the `recoverFromSleep()` function.